rfcs/20250312-ukernel-api-updates/README.md

```diff
@@ -0,0 +1,262 @@
+    /// @param C Pointer to a tensor C (accumulation buffer).
+    /// @param scratchpad Pointer to a scratchpad buffer.
+    /// @param actual_lds NEW ARGUMENT. Vector of actual leading dimensions for
+    ///     tensors A, B, and C. Must be specified if brgemm object was created
+    ///     with `DNNL_RUNTIME_DIM_VAL` instead of regular values for `lda`,
+    ///     `ldb`, or `ldc`. When `get_B_pack_type` returns values other than
+    ///     `no_trans`, `actual_lds[1]` for `ldb` must be used in transform
+    ///     routine. If any of given values were not specified as runtime values
+    ///     at kernel creation, they will be used instead.
+    void execute(const void *A, const void *B,
+            const std::vector<std::pair<memory::dim, memory::dim>> &A_B_offsets,
+            void *C, void *scratchpad,
+            const std::vector<memory::dim> &actual_lds = {}) const {
+        dnnl_status_t status = dnnl_brgemm_execute(get(), A, B,
+                (const dnnl_dim_t *)A_B_offsets.data(), C, scratchpad,
+                static_cast<const dnnl_dim_t *>(actual_lds.data()));
+        if (status != dnnl_success)
+            error::wrap_c_api(
+                    status, "could not execute a BRGeMM ukernel object");
+    }
+
+    /// ...
+    void execute(const void *A, const void *B,
+            const std::vector<std::pair<memory::dim, memory::dim>> &A_B_offsets,
+            const void *C, void *D, void *scratchpad,
+            const attr_params &params = default_attr_params(),
+            const std::vector<memory::dim> &actual_lds = {}) const {
+        dnnl_status_t status = dnnl_brgemm_execute_postops(get(), A, B,
+                (const dnnl_dim_t *)A_B_offsets.data(), C, D, scratchpad,
+                params.get(),
+                static_cast<const dnnl_dim_t *>(actual_lds.data()));
+        if (status != dnnl_success)
+            error::wrap_c_api(
+                    status, "could not execute a BRGeMM ukernel object");
+    }
+
+    ...
+};
+
+/// Transform ukernel
+struct transform : public handle<dnnl_transform_t> {
+    ...
+
+    /// Executes a transform object.
+    ///
+    /// @param in Pointer to an input buffer.
+    /// @param out Pointer to an output buffer.
+    /// @param in_actual_ld Actual input buffer leading dimension. Must be
+    ///     specified if transform was created with runtime `in_ld`.
+    void execute(
+            const void *in, void *out, memory::dim in_actual_ld = 0) const {
+        dnnl_status_t status
+                = dnnl_transform_execute(get(), in, out, in_actual_ld);
+        if (status != dnnl_success)
+            error::wrap_c_api(status,
+                    "could not execute a BRGeMM ukernel packing B object");
+    }
+
+    ...
+};
+```
+
+Usage:
+
+```cpp
+    ...
+
+    using namespace dnnl::ukernel;
+
+    const auto pack_type = brgemm::get_B_pack_type(a_dt, b_dt);
+    const bool need_pack = pack_type != pack_type::no_trans;
+
+    if (need_pack) {
```

Comment:
You haven't checked for `pack_type::undef`, although I'm guessing this will be handled by `transform` returning a null pointer. If that is the case, it might make sense to make this a documented and expected feature when implementing transforms.